libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectExTest.kt

```diff
@@ -0,0 +1,141 @@
+import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFromRoadObject
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.mockkObject
+import io.mockk.unmockkObject
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+
+class RoadObjectExTest {
+
+    @Before
+    fun setUp() {
+        mockkObject(RestStopFactory)
+    }
+
+    @After
+    fun tearDown() {
+        unmockkObject(RestStopFactory)
+    }
+
+    @Test
+    fun `get all rest stops using road object`() {
+        val upcomingRoadObjects = listOf<UpcomingRoadObject>(mockk(), mockk())
+        val mockRestStops = listOf<RestStopFromRoadObject>(mockk())
+        every { RestStopFactory.createWithUpcomingRoadObjects(any()) } returns mockRestStops
+
+        val actual = upcomingRoadObjects.getAllRestStops()
+
+        assertEquals(mockRestStops, actual)
+    }
+
+    @Test
+    fun `get first upcoming rest stop`() {
+        val upcomingRoadObjects = listOf<UpcomingRoadObject>(mockk(), mockk())
+        val mockRestStops = listOf<RestStopFromRoadObject>(
+            mockk {
+                every { distanceToStart } returns 123.0
+            },
+            mockk()
+        )
+
+        every { RestStopFactory.createWithUpcomingRoadObjects(any()) } returns mockRestStops
+
+        val actual = upcomingRoadObjects.getFirstUpcomingRoadObject()
+
+        assertEquals(actual, mockRestStops.first())
+    }
+
+    @Test
+    fun `get first upcoming rest stop when distanceToStart does not satisfy`() {
+        val upcomingRoadObjects = listOf<UpcomingRoadObject>(mockk(), mockk())
+        val mockRestStops = listOf<RestStopFromRoadObject>(
+            mockk {
+                every { distanceToStart } returns null
+            },
+            mockk()
+        )
+
+        every { RestStopFactory.createWithUpcomingRoadObjects(any()) } returns mockRestStops
+
+        val actual = upcomingRoadObjects.getFirstUpcomingRoadObject()
+
+        assertNull(actual)
+    }
+
+    @Test
+    fun `get first upcoming rest stop when empty`() {
+        val upcomingRoadObjects = listOf<UpcomingRoadObject>(mockk(), mockk())
+        val mockRestStops = emptyList<RestStopFromRoadObject>()
+
+        every { RestStopFactory.createWithUpcomingRoadObjects(any()) } returns mockRestStops
+
+        val actual = upcomingRoadObjects.getFirstUpcomingRoadObject()
+
+        assertNull(actual)
+    }
+
+    @Test
+    fun `no rest stops returned when route leg is null`() {
+        val navigationRoute = mockk<NavigationRoute> {
+            every { directionsRoute } returns mockk {
+                every { legs() } returns null
+            }
+        }
+
+        val actual = navigationRoute.getAllRestStops()
+
+        assertTrue(actual.isEmpty())
+    }
+
+    @Test
+    fun `no rest stops returned when route leg step is null`() {
```

Comment:
You can also check that other legs are not affected (only the one with null steps is ignored).